remove replicaset unit tests that are converted to integration tests

crimsonfaith91 opened the pull request:
**What this PR does / why we need it**:
This PR revamps existing replicaset unit tests by removing the tests that have been converted to integration tests.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: xref #52118

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

crimsonfaith91 commented:
/assign @enisoc @kow3ns

crimsonfaith91 commented:
@enisoc @kow3ns 
IMO, following tests should be retained as unit tests.

**TestGetCondition**, **TestSetCondition**, **TestRemoveCondition**
solid unit tests

**TestPodControllerLookup**
getting the list of matching controllers for a pod is not a behavior that should be externally observable

**TestControllerBurstReplicas**
would be racing with the real API server in integration test environment

**TestDoNotAdoptOrCreateIfBeingDeleted**
unsuitable as an integration test as it tests against something does not happen

**TestDoNotAdoptOrCreateIfBeingDeletedRace**
the test involves race condition, which is not feasible in the integration test environment

**TestSyncReplicaSetDoesNothing**
unsuitable as an integration test as it tests against something does not happen

**TestStatusUpdatesWithoutReplicasChange**
unsuitable as an integration test as it tests against something does not happen

**TestPatchPodFails**
unsuitable as an integration test as it requires a failure to be happened

**TestSyncReplicaSetDormancy**
unsuitable as an integration test as it requires a failure to be happened

**TestDeleteFinalStateUnknown**
the test verifies that a DeletedFinalStateUnknown object does cause the controller of the deleted pod to be inserted into work queue, which is not a behavior that should be externally observable

**TestRSSyncExpectations**
the test verifies that a pod cannot sneak in between counting active pods and checking expectations, which is not a behavior that should be externally observable

**TestDeleteControllerAndExpectations**
unsuitable as an integration test as it tests against something does not happen

**TestControllerUpdateRequeue**
the test verifies that a controller is forced to be re-queued after an update failure, which is not a behavior that should be externally observable

**TestWatchControllers**
the test examines how watcher, work queue and sync handler work with each other, which is not a behavior that should be externally observable

**TestWatchPods**
the test examines how watcher and sync handler work with each other, which is not a behavior that should be externally observable

**TestControllerUpdateStatusWithFailure**
unsuitable as an integration test as it requires a failure to be happened (e.g., verifies that GET is for the right RS when UPDATE failed)

**TestDeletionTimestamp**
the test verifies delete handler works correctly when a pod is set with a deletion timestamp (e.g., a pod with a non-nil deletion timestamp should be ignored by the delete handler), which is not a behavior that should be externally observable

enisoc commented:
Thanks for the detailed analysis! I agree with all your retained tests except one:

**TestStatusUpdatesWithoutReplicasChange**

Actually, you are testing this already in your integration test PR.
It's the one where you will update Spec to increment Generation,
and check that ObservedGeneration changes even though no other
changes in Status are needed.

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -248,24 +225,6 @@ func TestSyncReplicaSetDoesNothing(t *testing.T) {
 	validateSyncReplicaSet(t, &fakePodControl, 0, 0, 0)
 }
 
-func TestSyncReplicaSetDeletes(t *testing.T) {
```
Make sure your integration tests contain a step where you ensure N replicas, and then scale down to <N.

crimsonfaith91 replied:
sure, i will add the step.

crimsonfaith91 replied:
This step should have been added by: 
```
scaleRS(t, c, rs, 3)
scaleRS(t, c, rs, 0)
scaleRS(t, c, rs, 2)
```

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -395,53 +331,6 @@ func TestStatusUpdatesWithoutReplicasChange(t *testing.T) {
-	manager.syncReplicaSet(getKey(rs, t))
-
-	// 1. Status.Replicas should go up from 2->4 even though we created 5-4=1 pod.
-	// 2. Status.FullyLabeledReplicas should equal to the number of pods that
```
Do you have an integration test that checks FullyLabeledReplicas when it is not the same as Replicas?

crimsonfaith91 replied:
no, let me add it. thanks for catching it!

crimsonfaith91 replied:
@enisoc I managed to get all tests passed except this. In the test, I created a RS with selectors and labels as `map[extraKey:extraValue foo:bar]`. After the RS created its own pods, I changed one of the pod to have labels `map[foo:bar]`. 

I was able to verify the labels changed in the pooling loop:
```
before:  map[extraKey:extraValue foo:bar] map[extraKey:extraValue foo:bar]
after: map[foo:bar] map[extraKey:extraValue foo:bar]
after: map[foo:bar] map[extraKey:extraValue foo:bar]
...
```
However, `.Status.FullyLabeledReplicas` did not change (stayed at 2). Does the test algorithm looks correct to you?

enisoc replied:
The test should look like this:

* Create RS with only labels {foo: bar} and 2 replicas. Wait for it to be stable.
* Change RS Pod template labels to {foo: bar, extra: extra}. But *don't* change the selector.
* Change the labels on only 1 Pod to {foo: bar, extra: extra}.
* Wait for RS Status to show Replicas=2, FullyLabeledReplicas=1.

crimsonfaith91 replied:
Thanks! I verify the algorithm is working and the test is passing. :)
I never thought about don't change the selector - the selector has to match template labels for apps/v1beta2, but then I realized the testing framework still uses extensions/v1beta1. I will add a note about the test may break if the testing framework migrates to use newer API version.

enisoc replied:
The test shouldn't break for newer API versions. We require the selector to match the template in all versions.

When we say "match" we don't mean the set of labels has to be equal. We mean the selector will select objects with those labels.

For example, a selector of `{foo: bar}` "matches" an object with labels `{foo: bar, extra: extra}`.

crimsonfaith91 replied:
I got it - it is fine for selector to "match" some of the labels, not necessarily all.

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -680,131 +564,6 @@ func TestWatchPods(t *testing.T) {
 	}
 }
 
-func TestUpdatePods(t *testing.T) {
```
Let's not remove this yet. It's not covered by integration. We need to refactor this into a unit test around a function that just outputs which RSes should be enqueued for a given Pod update event. Can you add it to the tracking issue for unit test refactors?

crimsonfaith91 replied:
sure!

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1115,60 +874,6 @@ func TestDeleteControllerAndExpectations(t *testing.T) {
-	return shuffled
-}
-
-func TestOverlappingRSs(t *testing.T) {
```
This should also stay as a unit test, but refactored around a function that just outputs which RSes to enqueue in response to a given addPod event. We can't observe if RSes get needlessly enqueued in an integration test.

crimsonfaith91 replied:
There is an integration test named `TestOverlappingRSs` for this, so i remove this test. You are definitely right that the integration test does not fully cover this unit test. There is a need to check which RSs are enqueued due to an event.

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1275,72 +980,6 @@ func setupManagerWithGCEnabled(stopCh chan struct{}, objs ...runtime.Object) (ma
 	return manager, fakePodControl, informers
 }
 
-func TestDoNotPatchPodWithOtherControlRef(t *testing.T) {
```
This should stay a unit test (it checks that we do nothing) but refactored.

crimsonfaith91 replied:
noted.

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1369,101 +1008,6 @@ func TestPatchPodFails(t *testing.T) {
 	}
 }
 
-func TestPatchExtraPodsThenDelete(t *testing.T) {
```
Do we have any integration test to replace this?

crimsonfaith91 replied:
@enisoc It is partially covered by deleting pod of `TestReplicaSetBasics` integration test. If they look irrelevant, we should add this test back. I have no idea how to test number of `PATCH` operations in integration test environment. Is it possible?

enisoc replied:
I don't see anything in `TestReplicaSetBasics` that would cover this. It looks like in that integration test, the test is the thing setting the Pod as pending deletion; then you check how RS reacts to this.

What this test does is wait until an RS has all the Pods it wants, and then give it another one that it will want to adopt. The expected behavior is that RS should adopt the Pod, but then realize it has too many. So the RS should choose one of the Pods (possibly the new one; possibly one of the old ones) and delete it, so that the total number of Pods comes back to the desired count.

crimsonfaith91 replied:
Ya, we should add an integration test for this. Perhaps we can verify whether number of ALL pods within the namespace matches the desired count.

crimsonfaith91 commented:
@enisoc Thanks for confirming! I will remove `TestStatusUpdatesWithoutReplicasChange` and this PR should be good to go. :)
I will make sure integration test PR merges first before this PR merges.

enisoc commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -1513,99 +1057,6 @@ func TestDoNotAdoptOrCreateIfBeingDeletedRace(t *testing.T) {
-	rs.Spec.MinReadySeconds = 15
-	informers.Extensions().V1beta1().ReplicaSets().Informer().GetIndexer().Add(rs)
-
-	// First pod becomes ready 20s ago
```
I think your integration test only checks the case when all Pods go Available at the same time. Can you modify it so that only some of them do, so we can see that it works even when AvailableReplicas != ReadyReplicas?

crimsonfaith91 replied:
yup, this will be much better.

crimsonfaith91 commented:
@enisoc Thanks for confirming! I will remove `TestStatusUpdatesWithoutReplicasChange` and add back all suggested tests. Then, this PR should be good to go. :)
I will make sure integration test PR merges first before this PR merges.

crimsonfaith91 commented:
@enisoc I missed out the parts of code refactoring. There are still some things to do before this PR is good to go. Sorry for the confusion.

crimsonfaith91 commented:
@enisoc 
Removed `TestStatusUpdatesWithoutReplicasChange`.
Added back `TestUpdatePods`, `TestOverlappingRSs`, and `TestDoNotPatchPodWithOtherControlRef`.
Will squash if everything looks good.

enisoc commented:
LGTM. Please squash and ask @janetkuo to /approve, but we should probably not /lgtm until #52067 merges.

crimsonfaith91 commented:
@enisoc Agreed that we should not merge this until #52067 merges.

crimsonfaith91 committed:
remove replicaset unit tests that are converted to integration tests

crimsonfaith91 commented:
@janetkuo Anthony will give LGTM to this PR when #52067 merges. Need your approval for this PR.

crimsonfaith91 commented:
/retest

janetkuo commented:
/approve

crimsonfaith91 commented:
/retest

enisoc commented:
/lgtm
